rotary/rules/polish.smk

```diff
@@ -15,37 +15,120 @@ rule prepare_medaka_polish_input:
     run:
-        source_relpath = os.path.relpath(str(input),os.path.dirname(str(output)))
-        os.symlink(source_relpath,str(output))
-
+        source_relpath = os.path.relpath(str(input),os.path.dirname(str(output.fasta)))
+        os.symlink(source_relpath,str(output.fasta))
 
-rule polish_medaka:
+rule map_long_reads_for_medaka:
     input:
         qc_long_reads="{sample}/qc/{sample}_qc_long.fastq.gz",
         contigs="{sample}/{step}/medaka_input/{sample}_input.fasta"
     output:
-        dir=directory("{sample}/{step}/medaka"),
-        contigs="{sample}/{step}/medaka/{sample}_consensus.fasta",
-        calls_to_draft=temp(multiext("{sample}/{step}/medaka/calls_to_draft", '.bam', '.bam.bai')),
-        consensus_probs=temp("{sample}/{step}/medaka/consensus_probs.hdf")
+        calls_to_draft_bam=temp('{sample}/{step}/medaka/calls_to_draft.bam'),
+        calls_to_draft_bam_index=temp('{sample}/{step}/medaka/calls_to_draft.bam.bai')
     conda:
         "../envs/medaka.yaml"
     log:
-        "{sample}/logs/{step}/medaka.log"
+        "{sample}/logs/{step}/map_long_reads_medaka.log"
+    benchmark:
+        "{sample}/benchmarks/{step}/map_long_reads_medaka.txt"
+    threads:
+        config.get("threads",1)
+    params:
+        output_base_name='{sample}/{step}/medaka/calls_to_draft'
+    shell:
+        """
+        mini_align -i {input.qc_long_reads} -r {input.contigs} -m -p {params.output_base_name} -t {threads} > {log} 2>&1
+        """
+
+checkpoint generate_contig_manifest:
+    input:
+        "{sample}/{step}/medaka_input/{sample}_input.fasta"
+    output:
+        contig_manifest="{sample}/{step}/medaka/{sample}_contigs.txt"
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/contig_names_medaka.log"
     benchmark:
         "{sample}/benchmarks/{step}/medaka.txt"
+    shell:
+        """
+        seqkit seq -n {input} > {output.contig_manifest} 2>> {log}
+        """
+
+rule polish_contig_medaka:
+    input:
+        calls_to_draft_bam='{sample}/{step}/medaka/calls_to_draft.bam',
+        calls_to_draft_bam_index='{sample}/{step}/medaka/calls_to_draft.bam.bai',
+        contig_manifest="{sample}/{step}/medaka/{sample}_contigs.txt"
+    output:
+        contig_polished=temp('{sample}/{step}/medaka/results/{sample}_{contig}.hd5')
+    conda:
+        "../envs/medaka.yaml"
+    log:
+        "{sample}/logs/{step}/medaka_{sample}_{contig}.log"
+    benchmark:
+        "{sample}/benchmarks/{step}/medaka_{sample}_{contig}.txt"
     params:
         medaka_model=config.get("medaka_model"),
-        batch_size=config.get("medaka_batch_size")
+        batch_size=config.get("medaka_batch_size"),
+        medaka_results_dir='{sample}/{step}/medaka/results',
+    threads:
+        2
+    shell:
+        """
+        medaka consensus {input.calls_to_draft_bam} \
+          {params.medaka_results_dir}/{wildcards.sample}_{wildcards.contig}.hd5 \
```

Comment:
Does this line need to be converted into an output variable to be recognized by the DAG (e.g., so the files can be deleted if the rule fails, etc.)?

Reply:
Or does it need to be outside the output section to work properly with checkpointing somehow?

Reply:
resolved